src/vvwe.rs

```diff
@@ -107,24 +108,43 @@ impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
         match op.happens_after() {
             // Dang! we have a happens after relation!
             Some(dot) => {
-                // Let's buffer this operation then.
+                // Have we already seen the operation we depend on?
                 if !self.saw_site_dot(&dot) {
-                    self.buffer.insert(dot, op);
-                    // and do nothing
-                    None
+                    // No, so buffer this operation until we see it
+                    self.buffer.entry(dot)
+                        .or_default()
+                        .push(op);
+                    vec![]
                 } else {
-                    Some(op)
+                    // Yes! Free this operation and all other operations which depend on this one
+                    self.remove_from_buffer(op)
                 }
             }
             None => {
-                // Ok so we're not causally constrained, but maybe we already saw an associated
-                // causal operation? If so let's just delete the pair
-                match self.buffer.remove(&op.dot()) {
-                    Some(_) => None, // we are dropping the dependent op! that can't be right
-                    None => Some(op),
-                }
+                // The operation we're processing doesn't depend on anything but maybe we've
+                // already seen operations which depend on this one? If so return those and free
+                // things transitively.
+                self.remove_from_buffer(op)
+            }
+        }
+    }
+
+    //TODO: better name
+    fn remove_from_buffer(&mut self, op: T) -> Vec<T> {
```

Comment:
the unfortunate side-effect of this generalization is that the worst case behavior for an operation becomes `O(buffer)` since in the worst case the operation we're processing could free up the whole buffer. In practice I highly doubt that this grows out of control.

Reply:
that's fine, we need to leave some work for others ;)